searx/webapp.py

```diff
@@ -141,11 +141,47 @@
    or (werkzeug_reloader and os.environ.get("WERKZEUG_RUN_MAIN") == "true"):
     search_initialize(enable_checker=True)
 
-babel = Babel(app)
+ui_locale_codes = [l.replace('_', '-') for l in settings['locales'].keys()]
+
+
+def get_locale():
+    if 'locale' in request.form\
+       and request.form['locale'] in settings['locales']:
+        # use locale from the form
+        locale = request.form['locale']
+        locale_source = 'form'
+    elif request.preferences.get_value('locale') != '':
+        # use locale from the preferences
+        locale = request.preferences.get_value('locale')
+        locale_source = 'preferences'
+    else:
+        # use local from the browser
+        locale = _get_browser_or_settings_language(request, ui_locale_codes)
+        locale = locale.replace('-', '_')
+        locale_source = 'browser'
+
+    # see _get_translations function
+    # and https://github.com/searx/searx/pull/1863
+    if locale == 'oc':
+        request.form['use-translation'] = 'oc'
+        locale = 'fr_FR'
+
+    logger.debug(
+        "%s uses locale `%s` from %s", urllib.parse.quote(request.url), locale, locale_source
+    )
+
+    return locale
+
+
+babel = Babel()
+if hasattr(babel, "localeselector"):
+    babel.init_app(app)
+    babel.localeselector(get_locale)
+else:
+    babel.init_app(app, locale_selector=get_locale)
```

Comment:
If you're changing the versions in requirements.txt to be fixed (i.e. `3.1.0` which throws the AttributeError), is it worth having this if-else?

I'd think you either keep this approach but cater for (include) a lower version in the requirements file or have a cleaner diff (mine was [this](https://github.com/arachne-threat-intel/searx/commit/96c3dff)).